week4/spiral/neongreen/Main.hs

```diff
@@ -0,0 +1,36 @@
+import Data.Foldable (for_)
+
+{-
+Take a spiral with e.g. size 9 and look at the empty space inside:
+
+    *********
+            *       xxxxxxxx
+    ******* *              x
+    *     * *        xxxxx x
+    * *** * *        x   x x
+    * * * * *        x x x x
+    * *   * *        x xxx x
+    * ***** *        x     x
+    *       *        xxxxxxx
+    *********
+
+Apart from the extra space in the beginning, it's a spiral with size 7.
+-}
+
+inSpiral :: Int -> (Int, Int) -> Bool
+inSpiral n (x,y) = or [
+  -- First line, last line, and last column are always filled
+  y == 0, y == n, x == n-1,
+  -- Only one element in the first column isn't filled
+  x == 0 && y /= 1,
+  -- Otherwise, the filled elements are ones that wouldn't be filled in a
+  -- smaller spiral
+  not (inSpiral (n-2) (x-1, y-1)) ]
```

Comment:
Interesting!!

It took me a while to understand how this terminates. Each time you take the next-smaller spiral, but significantly important is that implies alternating between the negative/positive space of the "origin" spiral. Negative means gaps. Positive means lines. Eventually you _must_ either:
* Find a smaller _negative_ spiral where there is a match
* Find a smaller _positive_ spiral where there is a match

Visually this is how I drew it:

```
expand...
    pos          neg         pos         neg         pos
or  [F, not or  [F, not or  [F, not or  [F, not or  [T, ]]]]]
...reduce
        T           F           T           F   
        

expand...
    pos            neg         pos         neg
or  [F, not or  [F, not or  [F, not or  [T, ]]]]
...reduce
        F           T           F        
```

There are several factors converging here to provide a solution, at least four:

1. First, the acute observation that this problem can be broken down into shrinking spirals alternating between neg/pos. Our aim is to find the given x/y (target) in the outer-edges of a spiral and then notice which case of spiral we are in (see below, very subtle).

2. the nested `not`s wherein their odd or even quantity determines the final outcome

3. combining the two ideas above, pos spirals are always odd numbered, neg spirals are always  even, even `not`s stay even, odd `not`s cancels out, hence signalling to-draw-or-not.

4. this solution _requires_ lazy evaluation because there is not terminal case but rather a `True` within `or [...]` will cause `or` to short circuit (not evaluate the rest of the lists elements). Proof: 
  ```
❯   or [True, undefined]
True

❯   or [False, undefined]
*** Exception: Prelude.undefined
```
5. The next smaller spiral implies a change in coordinate space such that we lose row 1 and column 1 AKA pull up one, pull left one. To maintain the correct target location we must likewise translate it. For some reason this took me a bit of time to visualize as well. I helped myself see it like so (bias toward clarifying the `y` changes):

  ```
    *********       xxxxxxxx     !*****
            *        !     x          *
    *!***** *        xxxxx x      *** *
    *     * *        x   x x      * * *
    * *** * *        x x x x      *   *
    * * * * *        x xxx x      *****
    * *   * *        x     x
    * ***** *        xxxxxxx
    *       *
    *********
```



So overall, my breath is taken away : D I look forward to a day where I can crush problems with this level of succinctness.